src/mrack/providers/provider.py

```diff
@@ -30,16 +37,116 @@ def name(self):
+        the servers was not successfully provisioned. If that happens it issues deletion
+        of all already provisioned resources.
+
+        Return list of information about provisioned servers.
+        """
+        if not logger:
+            logger = logging.getLogger(self.name)
+
+        logger.info("Validating hosts definitions")
+        await self.validate_hosts(hosts)
+        logger.info("Host definitions valid")
+
+        logger.info("Checking available resources")
+        can = await self.can_provision(hosts)
+        if not can:
+            raise ValidationError("Not enough resources to provision")
+        logger.info("Resource availability: OK")
+
+        started = datetime.now()
+
+        count = len(hosts)
+        logger.info(f"Issuing provisioning of {count} hosts")
+        create_servers = []
+        for req in hosts:
+            awaitable = self.create_server(req)
+            create_servers.append(awaitable)
+        create_resps = await asyncio.gather(*create_servers)
+        logger.info("Provisioning issued")
+
+        logger.info("Waiting for all hosts to be available")
+        wait_servers = []
+        for create_resp in create_resps:
+            awaitable = self.wait_till_provisioned(create_resp)
+            wait_servers.append(awaitable)
+
+        server_results = await asyncio.gather(*wait_servers)
+        provisioned = datetime.now()
+        provi_duration = provisioned - started
+
+        logger.info("All hosts reached provisioning final state (ACTIVE or ERROR)")
+        logger.info(f"Provisioning duration: {provi_duration}")
+
+        errors = [res for res in server_results if res["status"] == "ERROR"]
+        if errors:
+            logger.info("Some host did not start properly")
+            for err in errors:
+                self.print_basic_info(err)
+            logger.info("Given the error, will delete all hosts")
+            await self.delete_hosts(server_results, logger)
+            raise ProvisioningError(errors)
+
+        hosts = [self.to_host(srv) for srv in server_results]
+        for host in hosts:
+            logger.info(host)
+        return hosts
+
+    async def delete_host(self, host):
+        """Delete provisioned host."""
         raise NotImplementedError()
 
-    async def delete_hosts(self, provisioning_results):
-        """Delete provisioned hosts based on input from provision_hosts."""
+    async def delete_hosts(self, hosts, logger):
+        """Issue deletion of all servers based on previous results from provisioning."""
+        logger.info("Issuing deletion")
+        delete_servers = []
+        for host in hosts:
+            awaitable = self.delete_host(host)
+            delete_servers.append(awaitable)
+        results = await asyncio.gather(*delete_servers)
+        logger.info("All servers issued to be deleted")
+        return results
+
+    def _get_host_info_from_prov_result(self, prov_result):
```

Comment:
Seems to me that length of this method is longer than `to_host` in all/almost all providers. It seems cleaner to me to have in base class:
```
def to_host(self, provisioning_result):
        """Transform provisioning result into Host object."""
       raise NotImplementedError()
```

and have content of `_get_host_info_from_prov_result` in `to_host`:

Reply:
I agree, also it's odd to require a 'private' method to be implemented.

With this change, please, consider to group all methods that require implementation before the one implemented by the base class.

Reply:
Final solution is okay to me, one public method responsible for translation the data and we keep `def to_host` as the `Host` instances factory for now.